src/readii/feature_extraction.py

```diff
@@ -152,6 +158,165 @@ def singleRadiomicFeatureExtraction(
 	return idFeatureVector
 
 
+def featureExtraction(
+	ctSeriesID: str,
+	pdImageInfo: pd.DataFrame,
+	imageDirPath: Path,
+	pyradiomicsParamFilePath: Optional[str] = None,
+	roiNames: Optional[str] = None,
+	negativeControl: Optional[str] = None,
+	randomSeed: Optional[int] = None,
+	keep_running: bool = False,
+) -> List[Dict[str, Any]]:
+	"""Extract PyRadiomics features for all ROIs present in a CT.
+
+	Parameters
+	----------
+	ctSeriesID : str
+			The CT series identifier
+	pdImageInfo : pd.DataFrame
+			DataFrame containing image metadata
+	imageDirPath : Path
+			Base directory containing image data
+	pyradiomics_params_path : Optional[str]
+			Path to PyRadiomics parameters file
+	roiNames : Optional[str]
+			Name pattern for the ROIs
+	negativeControl : Optional[str]
+			Type of negative control to generate
+	random_seed : Optional[int]
+			Random seed for reproducibility
+	keep_running : bool
+			Whether to continue on error
+
+	Returns
+	-------
+	List[Dict[str, Any]]
+			List of dictionaries containing features for each ROI
+	"""
+	dataset_directory = Path(imageDirPath)
+	ctSeriesInfo = pdImageInfo.loc[pdImageInfo["series_CT"] == ctSeriesID]
+	patID = ctSeriesInfo.iloc[0]["patient_ID"]
+
+	# Set up logger for this patient and series
+	plogger = logger.bind(patientID=patID, series_CT=ctSeriesID)
+
+	plogger.info("Starting Feature Extraction")
+
+	# Get absolute path to CT image files
+	try:
+		ctDirPath = dataset_directory / ctSeriesInfo.iloc[0]["folder_CT"]
+
+		plogger.debug("Loading CT images", ctDirPath=ctDirPath)
+		# Load CT by passing in specific series to find in a directory
+		ctImage = read_dicom_series(path=ctDirPath.as_posix(), series_id=ctSeriesID)
+
+		# Get list of segmentations to iterate over
+		segSeriesIDList = ctSeriesInfo["series_seg"].unique()
+
+		plogger.debug(
+			f"Found {len(segSeriesIDList)} segmentations.", segSeriesIDList=segSeriesIDList
+		)
+
+		# Initialize dictionary to store radiomics data for each segmentation (image metadata + features)
+		ctAllData = []
+
+		# Loop over every segmentation associated with this CT - only loading CT once
+		for _, segSeriesID in enumerate(segSeriesIDList):
+			segSeriesInfo = ctSeriesInfo.loc[ctSeriesInfo["series_seg"] == segSeriesID]
+
+			if (
+				# Check that a single segmentation file is being processed
+				len(segSeriesInfo) > 1
+				# Check that if there are multiple rows that it's not due to a CT with subseries (this is fine, the whole series is loaded)
+				and not segSeriesInfo.duplicated(subset=["series_CT"], keep=False).all()
+			):
+				errmsg = "Some kind of duplication of segmentation and CT matches not being caught. Check seg_and_ct_dicom_list in radiogenomic_output."
```

Comment:
Error message has outdated output directory name.

```suggestion
				errmsg = "Some kind of duplication of segmentation and CT matches not being caught. Check seg_and_ct_dicom_list in readii_output."
```